src/module_1/module_1_meteo_api.py

```diff
@@ -1,7 +1,248 @@
+    :param params: Dictionary with the parameters to be passed to the API call.
+    :return: The API response if successful, None otherwise.
+    """
+    call_attempts = 0
+    backoff_time = 60
+    while call_attempts < MAX_CALL_ATTEMPTS:
+        try:
+            response = requests.get(API_URL, params=params)
+            if response.status_code == 200:
+                if validate_response(response.json()):
+                    return response
+            elif response.status_code == 429:
+                # Use the Retry-After header to determine the cool-off time
+                cool_off_time = response.headers.get("Retry-After", backoff_time)
+                logging.info(
+                    f"Rate limit exceeded. Waiting for {cool_off_time} seconds."
+                )
+                logging.info(response.text)
+                time.sleep(cool_off_time)
+                backoff_time *= 2
+            else:
+                raise Exception(
+                    f"API request failed with status code {response.status_code}"
+                )
+        except Exception as e:
+            logging.info(f"API request failed with exception {e}")
+            break
+        call_attempts += 1
+        logging.info(f"Call attempts: {call_attempts}")
+    logging.info("Max number of call attempts reached. Returning empty result")
+    return response
+
+
+def get_data_meteo_api(
+    city: str, start_date: str, end_date: str, variable: str, model: str
+) -> pd.DataFrame():
+    """
+    Fetches daily weather data for a specific city and variable from the Meteo API.
+    :param city: The name of the city to fetch the data for. Must be a key in the \
+    COORDINATES dictionary.
+    :param start_date: The start date for the data in the format 'YYYY-MM-DD'.
+    :param end_date: The end date for the data in the format 'YYYY-MM-DD'.
+    :param variable: The weather variable to fetch the data for.
+    :return: A DataFrame containing the fetched data.
+    :raises ValueError: If the specified city is not in the COORDINATES dictionary.
+    """
+    if city not in COORDINATES.keys():
+        raise ValueError("City not available")
+    params = {
+        "latitude": COORDINATES[city]["latitude"],
+        "longitude": COORDINATES[city]["longitude"],
+        "start_date": start_date,
+        "end_date": end_date,
+        "models": model,
+        "daily": variable,
+    }
+    data = call_api(params)
+    if data is None:
+        return None
+    daily_data = data.json()["daily"]
+    # Convert the daily data into a DataFrame
+    df = pd.DataFrame(daily_data)
+    # Add the city name as a column
+    df["city"] = city
+    df["time"] = pd.to_datetime(df["time"])
+    return df
+
+
+def calculate_mean_std(df: pd.DataFrame) -> pd.DataFrame:
+    """
+    Calculates the mean and standard deviation for each prefix in the \
+    DataFrame's columns.
+    :param df: The DataFrame to calculate the mean and standard deviation for.
```

Comment:
This tells hardly anything. How this dataframe looks like? Which columns do we expect?

Reply:
Especificar algo como esto?:
So, if the DataFrame has columns 'temperature_2m_mean_df0', 'temperature_2m_mean_df1', 'precipitation_sum_df0', 'precipitation_sum_df1', 'soil_moisture_0_to_10cm_mean_df0', and 'soil_moisture_0_to_10cm_mean_df1', the function will calculate the mean and standard deviation for the 'temperature_2m_mean', 'precipitation_sum', and 'soil_moisture_0_to_10cm_mean' prefixes. Once the mean and standard deviation are calculated, the original columns are dropped and the new columns are added to the DataFrame. Other columns without underscores are left untouched(i.e. 'city' and 'time')

Reply:
I normally do the following:

    `df: a nxm DataFrame containing data about X with columns [col1, col2, col3]  and whatever as indices`

Just by reading that I can now what to expect in terms of shape, content and naming